### PR TITLE
detect subscriber naming collisions intelligently to avoid class reloading issues

### PR DESCRIPTION
**Pull request by heythisisnate**
this is to get around the class reloading issue in development/test mode
bot review me @davidrunger

### CODE REVIEW

**Comment by davidrunger** [comment]
Not worth updating the changelog?

**Comment by davidrunger** [comment]
Also probably updating the README and/or [the error message](https://github.com/hired/reactor/blob/9ce4b7c59c4031cb433d7ff07ba20d7a57d4d5a4/lib/reactor/models/concerns/subscribable.rb#L63) to include a pointer to the `undef_subscribers!` method and `ActionDispatch::Reloader.to_cleanup` would probably be quite useful to our users?

On the other hand, I'm not sure how many people are using Reactor outside of Hired. And maybe we just want to get this bugfix in ASAP.

File: lib/reactor.rb [hunk 1]
```diff
@@ -44,6 +44,22 @@ def self.with_subscriber_enabled(klass)
     yield if block_given?
     disable_test_mode_subscriber klass
   end
+
+  # In Rails, this can be used to avoid class reloading issues in development/test mode.
+  # Add the following to an initializer:
+  #
+  #     ActionDispatch::Reloader.to_cleanup do
```
Comment (davidrunger) [inline lib/reactor.rb]: Trying this out, calling `undef_subscribers!` in `to_cleanup` doesn't seem to stop the `EventHandlerAlreadyDefined` errors, but they do seem to stop when using `to_prepare`. Is this different from your experience?

**Comment by heythisisnate** [comment]
I have an idea to make this automatic. Will update.

**Comment by heythisisnate** [comment]
@davidrunger please take another look. I _think_ this works.

**Comment by davidrunger** [comment]
Sorry for the delay in reviewing; I got called into a couple consecutive meetings.

As before, `to_cleanup` does not seem to work, but `to_prepare` does seem to.

**Comment by heythisisnate** [comment]
@davidrunger completely different approach. I changed the way Reactor internally keeps track of subscriber names, so naming conflicts can be determined without having to check for constant definition collisions.

**Pull request merged**